xarray/conventions.py

```diff
@@ -949,7 +949,7 @@ def decode_cf_variable(name, var, concat_characters=True, mask_and_scale=True,
 
     original_dtype = data.dtype
 
-    if concat_characters and data.dtype.kind == 'S':
+    if concat_characters and data.dtype.kind in ['U', 'S', 'O']:
```

Comment:
I don't think this fix is quite right. Both stacking characters and decoding bytes -> unicode only make sense if the variable has a NumPy `S` dtype (i.e., fixed length bytes).

We do need to fix this bug.... but I think the check below that uses `string_encoding` is actually in the wrong place. It makes sense to check `_FillValue` in writing/encoding, not reading/decoding.
https://github.com/pydata/xarray/pull/1803/files#r159131975